src/main/java/org/opensearch/security/user/UserService.java

```diff
@@ -0,0 +1,311 @@
+            }
+        } else { // Not a service account
+            final List<String> foundRestrictedContents = RESTRICTED_FROM_USERNAME.stream().filter(accountName::contains).collect(Collectors.toList());
+            if (!foundRestrictedContents.isEmpty()) {
+                final String restrictedContents = foundRestrictedContents.stream().map(s -> "'" + s + "'").collect(Collectors.joining(","));
+                throw new AccountCreateOrUpdateException(RESTRICTED_CHARACTER_USE_MESSAGE + restrictedContents);
+            }
+
+            // if password is set, it takes precedence over hash
+            final String plainTextPassword = securityJsonNode.get("password").asString();
+            final String origHash = securityJsonNode.get("hash").asString();
+            if (plainTextPassword != null && plainTextPassword.length() > 0) {
+                contentAsNode.remove("password");
+                contentAsNode.put("hash", hash(plainTextPassword.toCharArray()));
+            } else if (origHash != null && origHash.length() > 0) {
+                contentAsNode.remove("password");
+            } else if (plainTextPassword != null && plainTextPassword.isEmpty() && origHash == null) {
+                contentAsNode.remove("password");
+            }
+
+            final boolean userExisted = internalUsersConfiguration.exists(accountName);
+
+            // when updating an existing user password hash can be blank, which means no
+            // changes
+
+            // sanity checks, hash is mandatory for newly created users
+            if (!userExisted && securityJsonNode.get("hash").asString() == null) {
+                throw new AccountCreateOrUpdateException(NO_PASSWORD_OR_HASH_MESSAGE);
+            }
+
+            // for existing users, hash is optional
+            if (userExisted && securityJsonNode.get("hash").asString() == null) {
+                // sanity check, this should usually not happen
+                final String hash = ((Hashed) internalUsersConfiguration.getCEntry(accountName)).getHash();
+                if (hash == null || hash.length() == 0) {
+                    throw new AccountCreateOrUpdateException("Existing user " + accountName + " has no password, and no new password or hash was specified.");
+                }
+                contentAsNode.put("hash", hash);
+            }
+        }
+
+        internalUsersConfiguration.remove(accountName);
+        contentAsNode.remove("name");
+
+        internalUsersConfiguration.putCObject(accountName, DefaultObjectMapper.readTree(contentAsNode,  internalUsersConfiguration.getImplementingClass()));
+
+        if (!securityJsonNode.get("service").isNull() && securityJsonNode.get("service").asString() == "true") { // Internal users update the config as part
+            saveAndUpdateConfiguration(client, CType.INTERNALUSERS, internalUsersConfiguration);
+        }
```

Comment:
The config always needs to be updated or the change won't replicate to the other nodes

Reply:
This needs to be set like this for use with the internalUserApiAction.